Use `select` to disable building static binaries if `--cpu=darwin`.

jmillikin-stripe opened the pull request:
This change allows kubectl to be built on MacOS machines using
`bazel build //cmd/kubectl`.
    
Mac OS X doesn't support static binaries because it does not have a
stable syscall API. Userspace binaries are expected to dynamically
link against libcrt instead.
    
https://developer.apple.com/library/content/qa/qa1118/_index.html

```release-note
None
```

ixdy commented:
/assign
/ok-to-test

Building kubernetes with bazel on MacOS is known to be broken (#35359), so thanks for tackling this.

Once we migrate to gazelle (#47558) this should be better. I think gazelle generates similar `select` statements to what you've written here.

If this works (and all tests pass) I'm fine with these fixes for now, though.

jmillikin-stripe commented:
/test pull-kubernetes-federation-e2e-gce

jmillikin-stripe commented:
Ignoring pull-kubernetes-federation-e2e-gce for now: I'm pretty sure bugs in this PR wouldn't manifest as test failures in federation code.

pull-kubernetes-verify is more of a problem. It's failing because `gazel` doesn't generate the same BUILD files as what this PR would submit, which makes sense. Do I need to implement this change in Gazel instead?

ixdy commented:
Possibly, though we're in the process of deprecating `gazel` (now `kazel`) and moving to `gazelle`.

If it's a small enough change I'm willing to accept PRs, though. They should be submitted against the `kubernetes/repo-infra` repo now. (We aren't yet using that, but it's something I'm actively working on.)

mikedanese commented:
I added a commit that removes some unnecessary selects and removes automanaged tag from rules so CI passes. I'm a bit hesitant to recommend dropping automanaged though since the next person that touches these packages may run into trouble that they don't know how to fix. @ianthehat and @jayconrod might be interested in looking at a diff of what it takes to build kubectl on darwin.

ianthehat commented:
We have been testing gazelle against kubernetes recently, and Jay recently (yesterday) got to the point where the only major differences when it updated the files was the correct automatic insertion of many select statements...
I am hoping that very soon we can verify that everything works with the BUILD files updated by gazelle and then show you the diff to see what you think.

jayconrod commented:
As Ian said, the latest version of Gazelle mostly just moves platform-specific sources to `select`. There shouldn't be many other spurious changes, but there are a lot of `selects` added. It's a pretty large change, but it could be broken up into many smaller changes.

That said, bazelbuild/rules_go#614 is a proposal for Gazelle to put srcs and deps in a separate .bzl file in some declarative format. If that's what Kubernetes wants to do, we can wait for that and do one migration instead of two.

ixdy commented:
sorry for not following up on this.

#50407 will fix some of the issues, but we'll still need to have some logic around avoiding static linking on darwin. If you don't want to rebase to fix that, I'll probably follow up in the near term.

jmillikin-stripe commented:
I'm willing to rebase once the tree is in a good place. #50407 is currently touching over 2000 files, so I want to wait for that to land and things to settle down before I rebase.

ixdy commented:
rebase away! (I suspect there may still be a variety of darwin issues.)

jmillikin-stripe committed:
Use `select` to disable building static binaries if `--cpu=darwin`.

This change allows kubectl to be built on MacOS machines using
`bazel build //cmd/kubectl`.

Mac OS X doesn't support static binaries because it does not have a
stable syscall API. Userspace binaries are expected to dynamically
link against libcrt instead.

https://developer.apple.com/library/content/qa/qa1118/_index.html

jmillikin-stripe commented:
Rebased to HEAD. Most of the changes were handled by the kazel migration, only thing left to do is disable static linking of the final go_binary.

ixdy commented:
/lgtm

jmillikin-stripe commented:
@mikedanese (or any other OWNER), can you take a look?

ixdy commented:
/retest

mikedanese commented:
/approve

mikedanese commented:
/approve no-issue
